Fix ansible failures in a multi network environment

TimothyAsirJeyasing opened the pull request:
https://github.com/Tendrl/tendrl-ansible/issues/68
Signed-off-by: Timothy Asir J <tjeyasin@redhat.com>

TimothyAsirJeyasing committed:
Fix ansible failures in a multi network environment

https://github.com/Tendrl/tendrl-ansible/issues/68
Signed-off-by: Timothy Asir J <tjeyasin@redhat.com>

TimothyAsirJeyasing commented:
@mbukatov , @r0h4n , @nthomas-redhat , @shtripat , Please review

cybernth commented on site.yml.sample:
```diff
@@ -37,6 +37,10 @@
 
+    # Moreover you need to use ip address of interface where etcd will be
+    # listening, to configure etcd itself.
+    - set_fact: etcd_ip_address={{hostvars[inventory_hostname]['ansible_default_ipv4']["address"]}}
```
What will happen if ansible scripts executed from a different server?

mbukatov replied:
I agree, this looks error prone.

TimothyAsirJeyasing replied:
@nthomas-redhat it works perfectly even from different server. Tested.

TimothyAsirJeyasing replied:
However i will verify in another setup and update here!

TimothyAsirJeyasing replied:
verified!

cybernth replied:
If it works, I am good here.
@mbukatov , what you think?

mbukatov replied:
I don't think this makes sense: it defines new fact `etcd_ip_address` for all machines in ansible inventory file as an ip address of default ipv4 interface as detected by ansible. This means that value of `etcd_ip_address` fact will be different on each machine and the only host for which it could be correct (as it's not guaranteed that this default will be working, in a similar way how it's not guaranteed for the default which is there right now) is tendrl server machine.

See this example:

```
$ cat etcd-test.yml
---
- hosts: all
  user: root
  tasks:
    - set_fact: etcd_ip_address={{hostvars[inventory_hostname]['ansible_default_ipv4']["address"]}}
    - debug: msg="{{etcd_ip_address}}"

- hosts: tendrl-server
  user: root
  tasks:
    - debug: msg="{{etcd_ip_address}}"

- hosts: mbukatov-usm1-gl1.usmqe.lab.eng.brq.redhat.com
  user: root
  tasks:
    - debug: msg="{{etcd_ip_address}}"
```
```
$ ansible-playbook -i mbukatov-usm1.hosts etcd-test.yml
[DEPRECATION WARNING]: ANSIBLE_HOSTS option, The variable is misleading as it can be a list of hosts and/or paths to inventory sources . This feature will be removed in version 2.8. Deprecation warnings 
can be disabled by setting deprecation_warnings=False in ansible.cfg.

PLAY [all] ***************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************
Friday 16 February 2018  17:07:45 +0100 (0:00:00.089)       0:00:00.089 ******* 
ok: [mbukatov-usm1-server.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-client.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl4.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl1.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl3.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl2.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl5.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl6.usmqe.lab.eng.brq.redhat.com]

TASK [set_fact] **********************************************************************************************************************************************************************************************
Friday 16 February 2018  17:07:49 +0100 (0:00:03.979)       0:00:04.069 ******* 
ok: [mbukatov-usm1-client.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-server.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl1.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl2.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl3.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl4.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl5.usmqe.lab.eng.brq.redhat.com]
ok: [mbukatov-usm1-gl6.usmqe.lab.eng.brq.redhat.com]

TASK [debug] *************************************************************************************************************************************************************************************************
Friday 16 February 2018  17:07:49 +0100 (0:00:00.159)       0:00:04.228 ******* 
ok: [mbukatov-usm1-client.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.25"
}
ok: [mbukatov-usm1-gl1.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.58"
}
ok: [mbukatov-usm1-server.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.13"
}
ok: [mbukatov-usm1-gl2.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.60"
}
ok: [mbukatov-usm1-gl3.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.62"
}
ok: [mbukatov-usm1-gl4.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.64"
}
ok: [mbukatov-usm1-gl5.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.15"
}
ok: [mbukatov-usm1-gl6.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.35"
}

PLAY [tendrl-server] *****************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************
Friday 16 February 2018  17:07:49 +0100 (0:00:00.149)       0:00:04.378 ******* 
ok: [mbukatov-usm1-server.usmqe.lab.eng.brq.redhat.com]

TASK [debug] *************************************************************************************************************************************************************************************************
Friday 16 February 2018  17:07:50 +0100 (0:00:00.668)       0:00:05.047 ******* 
ok: [mbukatov-usm1-server.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.13"
}

PLAY [mbukatov-usm1-gl1.usmqe.lab.eng.brq.redhat.com] ********************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************
Friday 16 February 2018  17:07:50 +0100 (0:00:00.041)       0:00:05.088 ******* 
ok: [mbukatov-usm1-gl1.usmqe.lab.eng.brq.redhat.com]

TASK [debug] *************************************************************************************************************************************************************************************************
Friday 16 February 2018  17:07:51 +0100 (0:00:00.858)       0:00:05.947 ******* 
ok: [mbukatov-usm1-gl1.usmqe.lab.eng.brq.redhat.com] => {
    "msg": "10.37.169.58"
}

PLAY RECAP ***************************************************************************************************************************************************************************************************
mbukatov-usm1-client.usmqe.lab.eng.brq.redhat.com : ok=3    changed=0    unreachable=0    failed=0   
mbukatov-usm1-gl1.usmqe.lab.eng.brq.redhat.com : ok=5    changed=0    unreachable=0    failed=0   
mbukatov-usm1-gl2.usmqe.lab.eng.brq.redhat.com : ok=3    changed=0    unreachable=0    failed=0   
mbukatov-usm1-gl3.usmqe.lab.eng.brq.redhat.com : ok=3    changed=0    unreachable=0    failed=0   
mbukatov-usm1-gl4.usmqe.lab.eng.brq.redhat.com : ok=3    changed=0    unreachable=0    failed=0   
mbukatov-usm1-gl5.usmqe.lab.eng.brq.redhat.com : ok=3    changed=0    unreachable=0    failed=0   
mbukatov-usm1-gl6.usmqe.lab.eng.brq.redhat.com : ok=3    changed=0    unreachable=0    failed=0   
mbukatov-usm1-server.usmqe.lab.eng.brq.redhat.com : ok=5    changed=0    unreachable=0    failed=0
```

mbukatov commented:
I would suggest to address this issue by removing the default completely, as suggested (among other things) here: https://github.com/Tendrl/tendrl-ansible/issues/74

cybernth commented:
Make sense. Let us not take this in a hurry. Let us park it for milestone 3

TimothyAsirJeyasing commented:
I remember we are not using etcd_ip_address in nodes.
However, We can use python -c "import socket; print socket.gethostbyname({{hostvars[inventory_hostname]['groups']['tendrl-server']}}[0])" so that it can provide the ip correctly on node and server.

$ cat etcd-test.yml
---
- hosts: all
  user: root
  tasks:
  - shell: python -c "import socket; print socket.gethostbyname({{hostvars[inventory_hostname]['groups']['tendrl-server']}}[0])"
    register: serverip
  - set_fact: etcd_ip_address="{{serverip.stdout}}"

- hosts: tendrl-server
  user: root
  tasks:
  - debug: msg="server point of view etcd ip {{etcd_ip_address}}"

- hosts: gluster-servers
  user: root
  tasks:
  - debug: msg="node point of view etcd ip {{etcd_ip_address}}"

Results:-

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
ok: [tendrl.node1]
ok: [tendrl.server1]

TASK [command] *****************************************************************
changed: [tendrl.node1]
changed: [tendrl.server1]

TASK [set_fact] ****************************************************************
ok: [tendrl.node1]
ok: [tendrl.server1]

PLAY [tendrl-server] ***********************************************************

TASK [setup] *******************************************************************
ok: [tendrl.server1]

TASK [debug] *******************************************************************
ok: [tendrl.server1] => {
    "msg": "server point of view etcd ip 192.168.122.178"
}

PLAY [gluster-servers] *********************************************************

TASK [setup] *******************************************************************
ok: [tendrl.node1]

TASK [debug] *******************************************************************
ok: [tendrl.node1] => {
    "msg": "node point of view etcd ip 192.168.122.178"
}

mbukatov commented:
Based on today's _Discuss peding tendrl-nsible issues and plan_ meeting, we will address this problem by not providing any defaults in the playbook file, as described in https://github.com/Tendrl/tendrl-ansible/issues/74
